feat: use ansible to download ONNX files

esteve opened the pull request:
## Description

<!-- Write a brief description of this PR. -->
This PR uses Ansible to download ONNX files.

Fixes https://github.com/autowarefoundation/autoware.universe/issues/3137

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

esteve committed:
feat: use ansible to download ONNX files

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve commented:
@kenji-miyake I don't have much experience with Ansible, could you please have a look to make sure I'm on the right path? I'll add more tasks for more files. Thanks :pray:

esteve committed:
feat: add remaining ONNX files

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve commented:
@xmfcx @mitsudome-r this PR is now ready for review

xmfcx commented:
Related: https://github.com/orgs/autowarefoundation/discussions/3649

xmfcx commented:
Let the user pass the "autoware_data" folder directory to the ansible script. (ref: https://github.com/autowarefoundation/autoware/blob/main/setup-dev-env.sh ) (default: `~/autoware_data/` as suggested by @esteve )
Then ONNX files will be downloaded into:

```
.../autoware_data/package_name/map.pcd
.../autoware_data/package_name/map.josm ?
.../autoware_data/package_name/model1.onnx
```

Details in: https://github.com/orgs/autowarefoundation/discussions/3649#discussioncomment-6475228

Once this PR is tested and merged,

We can modify the launch files to point to these new files.

And we can proceed to remove existing cmake references.

esteve committed:
feat: add data_dir option to choose the location to store models

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve committed:
feat: update artifacts per package

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve committed:
fix: fix paths for downloaded artifacts

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve commented:
@xmfcx this PR is ready for review now, I've added an option to specify the directory to download the files to and updated the ansible scripts to cover all the artifacts.

xmfcx commented:
@lexavtanke could you review this please? If you have any questions, please don't hesitate to ask me or @esteve :pray:

lexavtanke commented:
Downloading part works good for ONNX files. But I have mention that it can be confusing for the user that he need to create the structure inside the `autoware_data` folder by itself. May be it is better to add it to the ansible script. Cause then people use it for preparing the env in is a bit annoying then you need to start process again and again after you meet the errors like this 
![Screenshot from 2023-09-12 00-54-51](https://github.com/autowarefoundation/autoware/assets/37497658/e8372d87-2753-46a1-8e7a-15b486af8f5f)
![Screenshot from 2023-09-12 00-31-18](https://github.com/autowarefoundation/autoware/assets/37497658/06a3051a-e0a5-4ea1-b026-7da4dc28c8be)

I also checked [updates in documentation](https://github.com/autowarefoundation/autoware-documentation/pull/345) and didn't find the folder preparation in the current version.

esteve commented:
@lexavtanke and I discussed this over Discord in private, I agree that this PR needs to be updated so that the folders for each package are created before downloading the models, `ansible` does not create folders automatically and needs to be told to create them beforehand.

esteve committed:
fix: create directories before downloading models

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve commented:
@lexavtanke I've updated the PR to make sure that the directories are created per package before downloading the models, can you have another look? Thanks.

lexavtanke commented on ansible/roles/onnx/tasks/main.yaml:
```diff
@@ -0,0 +1,298 @@
+  ansible.builtin.get_url:
+    url: https://awf.ml.dev.web.auto/perception/models/traffic_light_classifier/v2/traffic_light_classifier_mobilenetv2_batch_6.onnx
+    dest: "{{ data_dir }}/traffic_light_classifier/traffic_light_classifier_mobilenetv2_batch_6.onnx"
+    checksum: md5:47255a11bde479320d703f1f45db1242
```
In my case it breaks here: 
![Screenshot from 2023-09-15 15-31-01](https://github.com/autowarefoundation/autoware/assets/37497658/be55a742-1314-4e7a-9f93-150f268f06f8)
 This solves it 
```suggestion
    checksum: md5:28b408710bcb24f4cdd4d746301d4e78
```

lexavtanke commented on ansible/roles/onnx/tasks/main.yaml:
```diff
@@ -0,0 +1,298 @@
+  ansible.builtin.get_url:
+    url: https://awf.ml.dev.web.auto/perception/models/traffic_light_classifier/v2/traffic_light_classifier_efficientNet_b1_batch_6.onnx
+    dest: "{{ data_dir }}/traffic_light_classifier/traffic_light_classifier_efficientNet_b1_batch_6.onnx"
+    checksum: md5:28b408710bcb24f4cdd4d746301d4e78
```
In my case it breaks here and this solves it
```suggestion
    checksum: md5:378526d9aa9fc6705cf399f7b35b3053
```

lexavtanke commented on ansible/roles/onnx/tasks/main.yaml:
```diff
@@ -0,0 +1,298 @@
+  ansible.builtin.get_url:
+    url: https://awf.ml.dev.web.auto/perception/models/traffic_light_classifier/v2/lamp_labels.txt
+    dest: "{{ data_dir }}/traffic_light_classifier/lamp_labels.txt"
+    checksum: sha256:baac220ef86b70c94673bbf982121f7d43591b1768ac05e98f235b23772a9bed
```
In my case it breaks here: 
![Screenshot from 2023-09-15 15-10-06](https://github.com/autowarefoundation/autoware/assets/37497658/88825110-6983-4bab-a80e-451da85584b1)
This solves it
```suggestion
    checksum: sha256:1a5a49eeec5593963eab8d70f48b8a01bfb07e753e9688eb1510ad26e803579d
```

lexavtanke commented:
@esteve Everything works good with creating folders but I experienced weird things during downloading some files: 
 
![Screenshot from 2023-09-15 15-31-01](https://github.com/autowarefoundation/autoware/assets/37497658/cd070b99-c76d-41b7-bb16-7b8ded586c90)
![Screenshot from 2023-09-15 15-10-06](https://github.com/autowarefoundation/autoware/assets/37497658/334b1901-cd86-4451-9a7a-976287e5b11e)
Some checksums don't match in my case. I made suggestions in the review.

And now I noticed that ansible role named `onnx`. I think it can be confusing if it will be used to download artifacts for other packages. May be it is worth to change it to something like `get_artfacts` or just `artifacts`. What do you think?

esteve commented:
@lexavtanke

> May be it is worth to change it to something like get_artfacts or just artifacts. What do you think?

Good idea, `artifacts` sounds good to me.

esteve committed:
fix: rename onnx role to artifacts

Co-authored-by: Alexey Panferov <37497658+lexavtanke@users.noreply.github.com>
Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

lexavtanke commented on ansible/playbooks/universe.yaml:
```diff
@@ -49,3 +54,7 @@
     - role: autoware.dev_env.git_lfs
+
+    # ONNX files
+    - role: autoware.dev_env.onnx
```
```suggestion
    - role: autoware.dev_env.artifacts
```

esteve committed:
fix: add mode for downloaded files

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve committed:
fix: fqdn builtin module

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve committed:
fix: add mode for created directories

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve committed:
fix: move Jinja2 template to the end of 'name' to fix lint error

Signed-off-by: Esteve Fernandez <esteve.fernandez@tier4.jp>

esteve commented:
@lexavtanke I've fixed all the ansible lint issues, could you have a look this PR again? Thanks.

lexavtanke reviewed (approved):
@esteve LGTM, thank you.

lexavtanke commented:
@xmfcx I guess we need approval from you. Thanks

esteve commented:
@xmfcx thanks :slightly_smiling_face:
